File: python/0018_4Sum.py
# class Solution(object):
#     def fourSum(self, nums, target):
#         """
#         :type nums: List[int]
#         :type target: int
#         :rtype: List[List[int]]
#         """
#         res=[]
#         nums.sort()#从小到大

#         for i in range(len(nums)-3):
#             if i>0 and nums[i] ==nums[i-1]:
#                     continue
#             for j in range(i+1,len(nums)-2):
#                 if j>i+1 and nums[j] ==nums[j-1]:
#                     continue
                
#                 left,right=j+1,len(nums)-1
#                 sum_t=nums[i]+nums[j]
#                 while right>left:
#                     if sum_t+nums[left]+nums[right]==target:
#                         res.append([nums[i],nums[j],nums[left],nums[right]])
#                         while left < right and nums[left] == nums[left+1]:
#                             left += 1
#                         while left < right and nums[right] == nums[right-1]:
#                             right -= 1
#                         left += 1; right -= 1
#                     elif target-sum_t-nums[left]-nums[right]>0:
#                         left+=1
#                     else:
#                         right-=1

#         return res


class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """

        def findNsum(l, r, target, N, result, results):
            if r-l+1 < N or N < 2 or target < nums[l]*N or target > nums[r]*N:  # early termination
                return
            if N == 2: # two pointers solve sorted 2-sum problem
                while l < r:
                    s = nums[l] + nums[r]
                    if s == target:
                        results.append(result + [nums[l], nums[r]])
                        l += 1
                        while l < r and nums[l] == nums[l-1]:
                            l += 1
                    elif s < target:
                        l += 1
                    else:
                        r -= 1
            else: # recursively reduce N
                for i in range(l, r+1):
                    if i == l or (i > l and nums[i-1] != nums[i]):
                        findNsum(i+1, r, target-nums[i], N-1, result+[nums[i]], results)

        nums.sort()
        results = []
        findNsum(0, len(nums)-1, target, 4, [], results)
        return results

File: cpp/0018_4Sum.cpp
/*
因为有case
[1000000000,1000000000,1000000000,1000000000]
0
所以删去了一个判断也就是注释部分
*/
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        if(nums.empty())
            return res;

        sort(nums.begin(),nums.end());

        vector<int> cur;
        dfs_find(nums,nums.size(),0,target,4,cur,res);

        return res;
        
    }
    void dfs_find(vector<int> & nums, int len,int start,int target,int k,vector<int> &cur,vector<vector<int>> & res)
    {
        // if(target<k*nums.front() || target>k*nums.back())
        //     return;

        if(k==2)
        {
            int l=start,r=len-1;

            while(l<r)
            {
                int cur_sum=nums[l]+nums[r];

                if(cur_sum==target)
                {
                    cur.push_back(nums[l]);
                    cur.push_back(nums[r]);

                    vector<int> cur_copy(cur);
                    res.push_back(cur_copy);
                    cur.pop_back();
                    cur.pop_back();

                    l++, r--;
                    while(l<r && nums[l]==nums[l-1])
                        l++;
                    while(l<r && nums[r]==nums[r+1])
                        r--;
                }

                else if(cur_sum<target)
                    l++;
                else
                    r--;
            }

            return;
        }

        for(int i=start;i<=len-k;i++)
        {
            int ele=nums[i];
            if(i>start && nums[i-1]==ele)
                continue;

            cur.push_back(ele);
            dfs_find(nums,len,i+1,target-ele,k-1,cur,res);
            cur.pop_back();
        }
    }
};
